### PR TITLE
[MAT-805] [MAT-635] [MAT-840] MAT with HARP (Okta) Login/Logout

### PR DESCRIPTION
**Pull request by jkotanchik-SB**
### [MAT-805](https://jira.cms.gov/browse/MAT-805): Okta login JavaScript

- New Login page utilizing the Okta sign-in widget.
- New servlet to retrieve environment specific HARP properties required by sign-in widget.

### [MAT-635](https://jira.cms.gov/browse/MAT-635): Create MAT session upon Okta login

- MAT operates normally when logging in with HARP or Legacy.

- New set-up flow for retrieving and setting user details with HARP instead of legacy username and password:
      1. HARP Login page stores tokens in Browser's Local Storage.
      2a. `Mat.java` retrieves the tokens from Local Storage and calls `MatContext.setUserDetailsByHarpId` which flows through to the server and grabs user details needed to create the `LoginModel` (Note: `LoginModel` is used in both legacy and HARP flows).
     2b. Uses token to create new Spring `Authentication`. Services use the Spring `SecurityContext` with this `Authentication` object to determine the logged-in user.
     3. Returned `LoginModel` is set in `MatContext`.
     4. MAT loads as before.

- Sign Out also logouts of HARP.
   - Okta's logout must be called by the UserAgent. Added hidden GWT based Form to call the logout endpoint.

- Legacy login page to be removed at completion of epic [MAT-629](https://jira.cms.gov/browse/MAT-629).

### Other Updates:

- `pom.xml` Replaced Maven Repository icm.edu.pl with Maven Central.
- `web.xml` Corrected xml namespace.
- Refactored multiple Event and Click Handlers to use lambdas.

### CODE REVIEW

File: src/main/java/mat/client/Login.java [hunk 1]
```diff
@@ -58,127 +57,81 @@ public void onSuccess(final Void arg0) {
...
 		});
-		
-		MatContext.get().getEventBus().addHandler(BackToLoginPageEvent.TYPE, new BackToLoginPageEvent.Handler() {
-			
-			@Override
-			public void onLoginFailure(final BackToLoginPageEvent event) {
-				MatContext.get().redirectToHtmlPage(ClientConstants.HTML_LOGIN);
-			}
+
+		MatContext.get().getEventBus().addHandler(SuccessfulLoginEvent.TYPE, event -> {
```
Comment (serhii-ilin) [inline src/main/java/mat/client/Login.java:93]: Does this actually work? I recall we had issues in similar cases. Java closure event handlers were not working properly and I had to rollback to 'Anonymous Inner Classes'..

Comment (cday-semanticbits) [reply src/main/java/mat/client/Login.java:93]: they are used all over mat gwt code now.

File: src/main/java/mat/server/PreventCachingFilter.java [hunk 2]
```diff
@@ -62,18 +62,19 @@ else if(requestURI.indexOf("/Mat.html") >= 0 || requestURI.indexOf("/Bonnie.html
...
-				httpResponse.setStatus(302);
-				String url = "Login.html";
-				if(httpRequest.getQueryString() != null) {
-					url = url + "?" + httpRequest.getQueryString();
-				}
-				httpResponse.setHeader("Location", url);
-			}
-			else {
+			//TODO Check if we need to do ALL of this no-cache and re-direct stuff.
+//			if(LoggedInUserUtil.getLoggedInUser() == null) {
```
Comment (cday-semanticbits) [inline src/main/java/mat/server/PreventCachingFilter.java:66]: TODO left open

File: src/main/java/mat/server/LoginServiceImpl.java [hunk 3]
```diff
@@ -111,6 +107,16 @@ public LoginModel isValidUser(String userId, String password, String oneTimePass
 		LoginModel loginModel = loginCredentialService.isValidUser(userId, password, oneTimePassword,session.getId());
 		return loginModel;
 	}
+
+	@Override
+	public LoginModel getUserDetailsByHarpId(String harpId, String accessToken) {
+		logger.info("getUserDetailsByHarpId::harpId::" + harpId);
+		HttpSession session = getThreadLocalRequest().getSession();
+//		if(session.getAttribute("accessToken") == null ) { //FIXME Attribute currently unused. Might be useful to determine the token type.
```
Comment (cday-semanticbits) [inline src/main/java/mat/server/LoginServiceImpl.java:115]: maybe remove this commented out code?

File: src/main/java/mat/server/PreventCachingFilter.java (hunk 2 above)
Comment (jkotanchik-SB) [reply src/main/java/mat/server/PreventCachingFilter.java]: Thanks. Updated TODO with jira ticket as it is not part of this task.

File: src/main/java/mat/server/LoginServiceImpl.java (hunk 3 above)
Comment (jkotanchik-SB) [reply src/main/java/mat/server/LoginServiceImpl.java]: Done. Never found a solid use case for it.

File: src/main/java/mat/server/service/impl/LoginCredentialServiceImpl.java [hunk 4]
```diff
@@ -330,6 +328,23 @@ public boolean isValidPassword(String userId, String password) {
         }
     }
 
+    public LoginModel initSession(String harpId, String sessionId, String accessToken) {
+        logger.debug("setUpUserSession::" + harpId + "::" + sessionId);
+        // Set Session ID in user details.
+        MatUserDetails userDetails = (MatUserDetails) hibernateUserService.loadUserByHarpId(harpId);
+        if(userDetails == null) {
+            throw new IllegalArgumentException("HARP_ID_NOT_FOUND");
```
Comment (RohitKandimalla) [inline src/main/java/mat/server/service/impl/LoginCredentialServiceImpl.java:336]: I have a new interface to store HarpConstants. We can use it for Harp related messages and Constants.

File: src/main/java/mat/server/model/MatUserDetails.java [hunk 5]
```diff
@@ -136,7 +136,7 @@ public void setUsername(String username) {
 	 */
 	@Transient
 	public Collection<GrantedAuthority> getAuthorities() {
-	       List<GrantedAuthority> list = new ArrayList<GrantedAuthority>();
+	       List<GrantedAuthority> list = new ArrayList<>();
```
Comment (RohitKandimalla) [inline src/main/java/mat/server/model/MatUserDetails.java:139]: Can we name this list specific to the context.

File: src/main/java/mat/server/PreventCachingFilter.java [hunk 6]
```diff
@@ -50,7 +50,9 @@ public void doFilter(ServletRequest request, ServletResponse response,
 			}
 			httpResponse.setHeader("Location", url);
 		}
-		else if(requestURI.indexOf("/Mat.html") >= 0 || requestURI.indexOf("/Bonnie.html") >= 0) {
+		//TODO MAT-864
+//		else if(requestURI.contains("/Mat.html") || requestURI.contains("/Bonnie.html")) {
```
Comment (RohitKandimalla) [inline src/main/java/mat/server/PreventCachingFilter.java:54]: Do we need this in the future ?

**Pull request merged**